src/AutoRest.CSharp/Common/Output/Models/Types/SignatureType.cs

```diff
@@ -0,0 +1,318 @@
+using AutoRest.CSharp.Output.Models.Shared;
+using Microsoft.CodeAnalysis;
+using static AutoRest.CSharp.Common.Output.Models.Snippets;
+
+namespace AutoRest.CSharp.Output.Models.Types
+{
+    /// <summary>
+    /// This type holds three portions of codes:
+    ///     - current
+    ///     - custom
+    ///     - baseline contract
+    ///     current union custom compare with baseline contract outputs the changeset, we can apply different rules with it.
+    /// </summary>
+    internal class SignatureType
+    {
+        private readonly TypeFactory _typeFactory;
+        private readonly string _defaultNamespace;
+        private readonly string _defaultName;
+        private readonly SignatureType? _customization;
+        private readonly SignatureType? _baselineContract;
+        private readonly MethodChangeset? _methodChangeset;
+
+        // Missing means the method with the same name is missing from the current contract
+        // Updated means the method with the same name is updated in the current contract, and the list contains the previous method and current methods including overload ones
+        private record MethodChangeset(IReadOnlyList<MethodSignature> Missing, IReadOnlyList<(List<MethodSignature> Current, MethodSignature Previous)> Updated) { }
+
+        public SignatureType(TypeFactory typeFactory, IReadOnlyList<MethodSignature> methods, SourceInputModel? sourceInputModel, string defaultNamespace, string defaultName)
+        {
+            // This can only be used for Mgmt now, because there are custom/hand-written code in HLC can't be loaded into CsharpType such as generic methods
+            if (!Configuration.AzureArm)
+            {
+                throw new InvalidOperationException("This type should only be used for Mgmt");
+            }
+
+            _typeFactory = typeFactory;
+            Methods = methods;
+            _defaultNamespace = defaultNamespace;
+            _defaultName = defaultName;
+            if (sourceInputModel is not null)
+            {
+                _customization = new SignatureType(typeFactory, PopulateMethodsFromCompilation(sourceInputModel?.Customization), null, defaultNamespace, defaultName);
+                _baselineContract = new SignatureType(typeFactory, PopulateMethodsFromCompilation(sourceInputModel?.PreviousContract), null, defaultNamespace, defaultName);
+                _methodChangeset ??= CompareMethods(Methods.Union(_customization?.Methods ?? Array.Empty<MethodSignature>(), MethodSignature.ParameterAndReturnTypeEqualityComparer), _baselineContract?.Methods);
+            }
+        }
+
+        private IReadOnlyList<Method>? _overloadMethods;
+        public IReadOnlyList<Method> OverloadMethods => _overloadMethods ??= EnsureOverloadMethods();
+
+        private IReadOnlyList<Method> EnsureOverloadMethods()
+        {
+            var overloadMethods = new List<Method>();
+            var updated = _methodChangeset?.Updated;
+            if (updated is null)
```

Comment:
nit: we could have:
```suggestion
            if (_methodChangeset is not {Updated: {} updated})
```
or
```suggestion
            if (_methodChangeset?.Updated is not {} updated)
```

Reply:
updated